### DISCUSSION TITLE
Colin McDonnell: tRPC, Zod, EdgeDB

### DISCUSSION THREAD

**Discussion by hipstersmoothie** [devtools-fm/devtools.fm#78] [Announcements]
## Intro

Andrew: Hello, welcome to the devtools.fm podcast. This is a podcast about developer tools and the people who make them. I'm Andrew and this is my co-host Justin.

Justin: Hey everyone! Today's guest is Colin McDonnell. Colin is the head of developer relations at EdgeDB and the creator of Zod, a typescript validation library. Thanks for joining us Colin! Would you like to tell our listeners a bit more about yourself?

## Questions

- [x] What you're background? What was your YC idea?
- [x] What is zod? What problem were you trying to solve with it?
- [x] Did you work on zod inspire you to create tRPC?
- [x] What is it good for what is it not?
- [x] Are there any cool examples of uses of zod you can think of? (outside typing API responses)
- [x] You say you should have just contributed to another library, why do you feel that way?
- [x] Did the name come from the superman villian?
- [x] What is EdgeDB and how does it differ from traditional relational databases?
- [x] You work on the query language for EdgeDB. What parts have you helped to improve? What type magic have you baked in?
- [x] Do you write edgeql directly from your typescript code? or do you use some intermediary?
- [x] What is feel like sitting atop that TypeScript generic throne? But in all seriousness how did you get so good at writing this type of code?

## Outro

Andrew: Well that’s it for this week’s episode of devtools.fm. Be sure to follow us on YouTube and wherever you consume your podcasts. Thanks for listening!

**Comment by hipstersmoothie** [comment]
## Research

Background
- CS at MIT
- Digital nomad 2016-2020
- YC W19
- Tried building medical software with Neo4j…bad idea

Zod
- Schema library
- Strong typing 
- Felt strongly that other libraries had usability issues
- Wrote a launch blog post
- Though in retrospect I probably should have just contributed to another library
- Happy to talk a bit about the general implementation approach and it’s strength/weaknesses

tRPC
- wrote initial version before handing it off to Alex
- [https://twitter.com/colinhacks/status/1344007284033114118…](https://twitter.com/colinhacks/status/1344007284033114118)

EdgeDB
- EdgeDB: an “object-relational” database
- declarative/object-oriented schema, migrations, and SQL 2.0
- i joined the company to build an EdgeQL query builder for Typescript
- auto-generated by introspecting your database schema
- write any query, full type inference, full type checking
- we basically implemented an entirely different type system inside of TypeScript

**Reply by just-be-dev** [reply]
We talked to Alex Johansson about tRPC back in episode 21

**Comment by hipstersmoothie** [comment]
My tooltips:

- https://github.com/reactjs/rfcs/pull/220

**Comment by just-be-dev** [comment]
- https://github.com/wycats/starbeam

**Comment by colinhacks** [comment]
https://www.amazon.com/Saddle-Chair/s?k=Saddle+Chair

https://www.npmjs.com/package/ts-node `-T` flag

https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/  ESM support
